---
title: "Selection analysis on Single Cell data"
author: "Alberto Vicens Sanchez"
date: "November 13, 2017"
output:
  html_document:
    df_print: paged
---

```{r setup, echo = FALSE, include=FALSE}
knitr::opts_knit$set(root.dir = "/home/user/phylogenomics_lab/sc_selection/")
knitr::opts_chunk$set(echo =FALSE, eval=FALSE, message = FALSE, warning=FALSE, cache = TRUE, fig.path = "/home/user/Dropbox/phylogenomics_lab_dbx/figures/", cache.path="/home/user/phylogenomics_lab/sc_selection/cache/")
```

In this project, we aim to test the evolutionary mechanisms driving the evolution of tumor populations through estimation of the **pN/pS** ratio, implying the number of non-synonymous mutations per non-synonymous site (pN) divided by the number of synonymous (silent) mutations per synonymous site (pS). We will try to estimate pN/pS in different branches and clades of phylogenetic tumor trees inferred from single-cell sequencing datasets.

#Single Cell Sequencing (SCS) data

We used publicly available whole-exome sequencing (WXS) datasets from three single-cell studies:

  * 1) *Wang et al. (2014). Clonal evolution in breast cancer revealed by single nucleus genome sequencing. Nature 512, 155–160:*
      55 Single cells derived from the primary tumor of a single breast cancer patient (ERBC). *W55 dataset*
  * 2) *Xu et al. (2012):*
      25 Single cells derived from the primary tumor of a single kidney tumor patient. *X25 dataset*
  * 3) *Hou et al. (2012) Cell:*
      65 Single cells derived from a single JAK-2 negative neoplasm myeloproliferative patient.

Genomic analysis and single nucleotide variant (SNVs) calling were performed following the pipeline of Alves & Posada (2018).

#Annotation of variants

We restrict this pipeline to the data from Wang et al. (2014) study but the very same pipeline was applied to the data from the remaining studies.

In SCS data, SNVs can be clasified as **clonal** (if they were also found in the bulk) or **subclonal** (if they were found in at least two tumor cells but NOT in the bulk). In W55 dataset, we obtained these SNVs:
  * 61 clonal mutations. VCF file: *W55.Clonal.vcf*
  * 69175 subclonal mutations. VCF file: *W55.Subclonal.vcf*

We annotated the variants using ANNOVAR (*http://annovar.openbioinformatics.org*) using the *refGene* library (gene-based annotation).

```{bash, eval=FALSE}
table_annovar.pl W55.Clonal.vcf humandb/ -buildver hg19 -out W55.Clonal -remove -protocol refGene -operation g -nastring . -vcfinput
table_annovar.pl W55.Subclonal.vcf humandb/ -buildver hg19 -out W55.Subclonal -remove -protocol refGene -operation g -nastring . -vcfinput
```

```{r annot_plot, eval=TRUE, dpi=100, fig.width=8, fig.height=6, fig.cap= "Figure1. Classification of Single Nucleotide Variants (SNVs) identified from W55 data. Barplots for Clonal and Subclonal datsets are shown. Variants were annotated with ANNOVAR." }
clonalvar<-read.table("data/W55.Clonal.Func.refGene.txt", header=T, sep ="\t")
subclonvar<-read.table("data/W55.Subclonal.Func.refGene.txt", header=T, sep ="\t")

cl_labels<-levels(clonalvar$Func.refGene)
sub_labels<-levels(subclonvar$Func.refGene)
layout(matrix(c(1,2),1,2), widths = c(3,7), heights = 10, respect = TRUE)

pl1<-plot(clonalvar$Func.refGene, ylim = c(0, 60), main= "W55 Clonal", ylab = "Number of Variants", xaxt="n", xlab="", cex.axis=0.7)
text(pl1,par("usr")[3], labels = cl_labels, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex=.7)
pl2<-plot(subclonvar$Func.refGene, ylim = c(0,35000), main= "W55 Subclonal", xaxt="n", xlab="", cex.axis=0.7)
text(pl2,par("usr")[3], labels = sub_labels, srt = 45, adj = c(1.1,1.1), xpd = TRUE, cex=.7)
``` 

We noted that the number of called variants is overestimated in the Subclonal dataset. This could be attributed to contamination of germline variants or an excessive false positive calling by MONOVAR.

#Processing of variant files
We prepare the mutant files to be used in the posterior analysis, following these several steps:

1)Filter for exonic mutations.
```{bash}
#Notice we take this step to remove the headers of the VCF file
grep -e '^#CHR' -e '=exonic' W55.Clonal.hg19_multianno.vcf > W55_Clonal_exonic_multianno.vcf
grep -e '^#CHR' -e '=exonic' W55.Subclonal.hg19_multianno.vcf > W55_Subclonal_exonic_multianno.vcf
```

2)Merge and sort the VCF files using the **vcf-concat** and **vcf-sort** tools from VCFtools. 
```{bash}
#Notice we take this step to remove the header generated by VCFtools.
vcf-concat  W55_Clonal_exonic_multianno.vcf W55_Subclonal_exonic_multianno.vcf  | vcf-sort | tail -n +2 > W55_All_exonic_multianno.vcf 
```
3)Rename the column headers of genotypes -names of cells-
```{bash}
sed 's/.sorted.Dedup.Real.Recal.bam//g' W55_All_exonic_multianno.vcf > tmp
mv tmp W55_All_exonic_multianno.vcf 
```

#Global pN/pS estimations
To estimate pN/pS, we used the **dNdScv** from *Martincorena I, et al. 2017*. This R package contains functions to quantify dN/dS ratios for missense, nonsense and essential splice mutations in cohorts of samples, at the level from individual to whole-genome level. We apply this method in order to estimate pN/pS on mutations identified from single cell sequencing.

```{bash convert_vcf_dNdScv}
#we need convert the VCF file to the input format of dNdScv. This is a data frame providing mutations with five columns (sampleID, chromosome, position, reference base and mutant base).
grep -v '^#' W55_All_exonic_multianno.vcf | cut -f1,2,4,5 | sed "s/^/W55\t/" | sed "1s/^/sample\tchr\tpos\tref\talt\n/" > W55_All_exonic_mut.txt
```

```{r run_dndscv1}
library("seqinr")
library("Biostrings")
library("MASS")
library("GenomicRanges")
library("dndscv")

#Load mutations file
w55mut<-read.table("/home/user/phylogenomics_lab/sc_selection/data/W55_All_exonic_mut.txt",header=T)
x25mut<-read.table("/home/user/phylogenomics_lab/sc_selection/data/X25_All_exonic_mut.txt",header=T)
w55mut_cl<-read.table("/home/user/phylogenomics_lab/sc_selection/data/W55_Clonal_exonic_mut.txt",header=T)
w55mut_subcl<-read.table("/home/user/phylogenomics_lab/sc_selection/data/W55_Subclonal_exonic_mut.txt",header=T)
x25mut_cl<-read.table("/home/user/phylogenomics_lab/sc_selection/data/X25_Clonal_exonic_mut.txt",header=T)
x25mut_subcl<-read.table("/home/user/phylogenomics_lab/sc_selection/data/X25_Subclonal_exonic_mut.txt",header=T)

#Run dNdScv
w55_dNdS<-dndscv(w55mut, outp = 1)
x25_dNdS<-dndscv(x25mut, outp = 1)
w55_cl_dNdS<-dndscv(w55mut_cl, outp =1)
w55_subcl_dNdS<-dndscv(w55mut_subcl, outp =1)
x25_cl_dNdS<-dndscv(x25mut_cl, outp =1)
x25_subcl_dNdS<-dndscv(x25mut_subcl, outp =1)
```


```{r global_dNdS_plot, dpi=100, fig.width=6, fig.height=6, fig.cap="Figure 2. Global pN/pS estimates for mutations obtained by whole exome sequencing data from single tumor cells"}
global_dnds<-rbind(w55_dNdS$globaldnds,x25_dNdS$globaldnds)
dataset<-rep(c("W55","X25"),each=5)
global_dnds <-cbind(dataset,global_dnds)

subset_global_dnds<-global_dnds[(global_dnds$name!= "wnon" &  global_dnds$name!= "wspl"),] #filter out nonsense and splicing mutations due to Inf estimates

ggplot(subset_global_dnds,aes(x=name,y=mle, colour=dataset)) + geom_errorbar(aes(ymin=cilow,ymax=cihigh),width=.2, 
                              position=position_dodge(.9)) + geom_point(position=position_dodge(0.9),size=2) + xlab("Mutation type") + ylab("Global dN/dS MLE [95% CI]") + ggtitle("Global dN/dS estimates - SCS data") +  scale_x_discrete(limit = c("wmis", "wtru","wall"),labels = c("Missense","Truncating", "All"))
```


```{r global_dNdS_plot_2, eval = TRUE, dpi=100, fig.width=6, fig.height=6, fig.cap="Figure 2. Overall pN/pS estimates for misense mutations obtained by whole exome sequencing data from single tumor cells. Estimates were compared among all, clonal (identified in bulk sequencing) and subclonal (not detected in bulk sequencing) mutation datsets"}
library("ggplot2")
#Load data frame of global dN/dS
ov_dnds<-read.table("data/sc_overall_dnds.txt",header = T)
#Filter for missense mutations
mis_dnds<-ov_dnds[(ov_dnds$name== "wmis"),]

ggplot(mis_dnds,aes(x=subtype,y=mle, colour=dataset)) + geom_errorbar(aes(ymin=cilow,ymax=cihigh),width=.2, position=position_dodge(.9)) + geom_point(position=position_dodge(0.9),size=2) + xlab("Mutation type") + ylab("Global pN/pS missense [95% CI]") + ggtitle("Global pN/pS estimates - SCS data")
```
#References
Martincorena I, et al. (2017) Universal Patterns of Selection in Cancer and Somatic Tissues. Cell. http://www.cell.com/cell/fulltext/S0092-8674(17)31136-4

Wang K, Li M, Hakonarson H. ANNOVAR: Functional annotation of genetic variants from next-generation sequencing data Nucleic Acids Research, 38:e164, 2010

Wang Y, Waters J, Leung ML, Unruh A, Rohet W et al. 2014. Clonal evolution in breast cancer revealed by single nucleus genome sequencingNature 512, 155–160. doi:10.1038/nature13600

Xu  X,  Hou  Y,  Yin  X,  Bao  L,  Tang  A,  Song  L,  et  al.  Single-cell  exome sequencing reveals single-nucleotide mutation characteristics of a kidney tumor.Cell. 2012;148: 886–895.
